File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-option-shared-5-4/AveragingPeriod.cpp
// AveragingPeriod.cpp 
#include "AveragingPeriod.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

AveragingPeriod::AveragingPeriod(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //scheduleNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* scheduleNode = xmlNode->FirstChildElement("schedule");

   if(scheduleNode){scheduleIsNull_ = false;}
   else{scheduleIsNull_ = true;}

   if(scheduleNode)
   {
      for(scheduleNode; scheduleNode; scheduleNode = scheduleNode->NextSiblingElement("schedule")){
          #ifdef ConsolePrint
              FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- scheduleNode , address : " << scheduleNode << std::endl;
          #endif
          if(scheduleNode->Attribute("href") || scheduleNode->Attribute("id"))
          {
              if(scheduleNode->Attribute("id"))
              {
                  scheduleIDRef_ = scheduleNode->Attribute("id");
                  schedule_.push_back(boost::shared_ptr<AveragingSchedule>(new AveragingSchedule(scheduleNode)));
                  schedule_.back()->setID(scheduleIDRef_);
                  IDManager::instance().setID(scheduleIDRef_, schedule_.back());
              }
              else if(scheduleNode->Attribute("href")) { scheduleIDRef_ = scheduleNode->Attribute("href");}
              else { QL_FAIL("id or href error"); }
          }
          else { schedule_.push_back(boost::shared_ptr<AveragingSchedule>(new AveragingSchedule(scheduleNode)));}
      }
   }
   else {
       #ifdef ConsolePrint
           FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- scheduleNode , address : " << scheduleNode << std::endl;
       #endif
   }

   //averagingDateTimesNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* averagingDateTimesNode = xmlNode->FirstChildElement("averagingDateTimes");

   if(averagingDateTimesNode){averagingDateTimesIsNull_ = false;}
   else{averagingDateTimesIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- averagingDateTimesNode , address : " << averagingDateTimesNode << std::endl;
   #endif
   if(averagingDateTimesNode)
   {
      if(averagingDateTimesNode->Attribute("href") || averagingDateTimesNode->Attribute("id"))
      {
          if(averagingDateTimesNode->Attribute("id"))
          {
             averagingDateTimesIDRef_ = averagingDateTimesNode->Attribute("id");
             averagingDateTimes_ = boost::shared_ptr<DateTimeList>(new DateTimeList(averagingDateTimesNode));
             averagingDateTimes_->setID(averagingDateTimesIDRef_);
             IDManager::instance().setID(averagingDateTimesIDRef_,averagingDateTimes_);
          }
          else if(averagingDateTimesNode->Attribute("href")) { averagingDateTimesIDRef_ = averagingDateTimesNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { averagingDateTimes_ = boost::shared_ptr<DateTimeList>(new DateTimeList(averagingDateTimesNode));}
   }

   //averagingObservationsNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* averagingObservationsNode = xmlNode->FirstChildElement("averagingObservations");

   if(averagingObservationsNode){averagingObservationsIsNull_ = false;}
   else{averagingObservationsIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- averagingObservationsNode , address : " << averagingObservationsNode << std::endl;
   #endif
   if(averagingObservationsNode)
   {
      if(averagingObservationsNode->Attribute("href") || averagingObservationsNode->Attribute("id"))
      {
          if(averagingObservationsNode->Attribute("id"))
          {
             averagingObservationsIDRef_ = averagingObservationsNode->Attribute("id");
             averagingObservations_ = boost::shared_ptr<AveragingObservationList>(new AveragingObservationList(averagingObservationsNode));
             averagingObservations_->setID(averagingObservationsIDRef_);
             IDManager::instance().setID(averagingObservationsIDRef_,averagingObservations_);
          }
          else if(averagingObservationsNode->Attribute("href")) { averagingObservationsIDRef_ = averagingObservationsNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { averagingObservations_ = boost::shared_ptr<AveragingObservationList>(new AveragingObservationList(averagingObservationsNode));}
   }

   //marketDisruptionNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* marketDisruptionNode = xmlNode->FirstChildElement("marketDisruption");

   if(marketDisruptionNode){marketDisruptionIsNull_ = false;}
   else{marketDisruptionIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- marketDisruptionNode , address : " << marketDisruptionNode << std::endl;
   #endif
   if(marketDisruptionNode)
   {
      if(marketDisruptionNode->Attribute("href") || marketDisruptionNode->Attribute("id"))
      {
          if(marketDisruptionNode->Attribute("id"))
          {
             marketDisruptionIDRef_ = marketDisruptionNode->Attribute("id");
             marketDisruption_ = boost::shared_ptr<MarketDisruption>(new MarketDisruption(marketDisruptionNode));
             marketDisruption_->setID(marketDisruptionIDRef_);
             IDManager::instance().setID(marketDisruptionIDRef_,marketDisruption_);
          }
          else if(marketDisruptionNode->Attribute("href")) { marketDisruptionIDRef_ = marketDisruptionNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { marketDisruption_ = boost::shared_ptr<MarketDisruption>(new MarketDisruption(marketDisruptionNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
std::vector<boost::shared_ptr<AveragingSchedule>> AveragingPeriod::getSchedule()
{
   if(!this->scheduleIsNull_){
        if(scheduleIDRef_ != ""){
             return this->schedule_;
        }else{
             return this->schedule_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return std::vector<boost::shared_ptr<AveragingSchedule>>();
   }
}
boost::shared_ptr<DateTimeList> AveragingPeriod::getAveragingDateTimes()
{
   if(!this->averagingDateTimesIsNull_){
        if(averagingDateTimesIDRef_ != ""){
             return boost::shared_static_cast<DateTimeList>(IDManager::instance().getID(averagingDateTimesIDRef_));
        }else{
             return this->averagingDateTimes_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<DateTimeList>();
   }
}
boost::shared_ptr<AveragingObservationList> AveragingPeriod::getAveragingObservations()
{
   if(!this->averagingObservationsIsNull_){
        if(averagingObservationsIDRef_ != ""){
             return boost::shared_static_cast<AveragingObservationList>(IDManager::instance().getID(averagingObservationsIDRef_));
        }else{
             return this->averagingObservations_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<AveragingObservationList>();
   }
}
boost::shared_ptr<MarketDisruption> AveragingPeriod::getMarketDisruption()
{
   if(!this->marketDisruptionIsNull_){
        if(marketDisruptionIDRef_ != ""){
             return boost::shared_static_cast<MarketDisruption>(IDManager::instance().getID(marketDisruptionIDRef_));
        }else{
             return this->marketDisruption_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<MarketDisruption>();
   }
}
}

File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-option-shared-5-4/AveragingPeriod.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class AveragingPeriod : ISerialized
    {
        public AveragingPeriod(XmlNode xmlNode)
        {
            XmlNodeList scheduleNodeList = xmlNode.SelectNodes("schedule");
            
            if (scheduleNodeList != null)
            {
                this.schedule_ = new List<AveragingSchedule>();
                foreach (XmlNode item in scheduleNodeList)
                {
                    if (item.Attributes["href"] != null || item.Attributes["id"] != null) 
                    {
                        if (item.Attributes["id"] != null) 
                        {
                            scheduleIDRef_ = item.Attributes["id"].Value;
                            schedule_.Add(new AveragingSchedule(item));
                            IDManager.SetID(scheduleIDRef_, schedule_[schedule_.Count - 1 ]);
                        }
                        else if (item.Attributes["href"] != null)
                        {
                            scheduleIDRef_ = item.Attributes["href"].Value;
                        }
                        else
                        {
                        schedule_.Add(new AveragingSchedule(item));
                        }
                    }
                    else
                    {
                        schedule_.Add(new AveragingSchedule(item));
                    }
                }
            }
            
        
            XmlNode averagingDateTimesNode = xmlNode.SelectSingleNode("averagingDateTimes");
            
            if (averagingDateTimesNode != null)
            {
                if (averagingDateTimesNode.Attributes["href"] != null || averagingDateTimesNode.Attributes["id"] != null) 
                {
                    if (averagingDateTimesNode.Attributes["id"] != null) 
                    {
                        averagingDateTimesIDRef_ = averagingDateTimesNode.Attributes["id"].Value;
                        DateTimeList ob = new DateTimeList(averagingDateTimesNode);
                        IDManager.SetID(averagingDateTimesIDRef_, ob);
                    }
                    else if (averagingDateTimesNode.Attributes["href"] != null)
                    {
                        averagingDateTimesIDRef_ = averagingDateTimesNode.Attributes["href"].Value;
                    }
                    else
                    {
                        averagingDateTimes_ = new DateTimeList(averagingDateTimesNode);
                    }
                }
                else
                {
                    averagingDateTimes_ = new DateTimeList(averagingDateTimesNode);
                }
            }
            
        
            XmlNode averagingObservationsNode = xmlNode.SelectSingleNode("averagingObservations");
            
            if (averagingObservationsNode != null)
            {
                if (averagingObservationsNode.Attributes["href"] != null || averagingObservationsNode.Attributes["id"] != null) 
                {
                    if (averagingObservationsNode.Attributes["id"] != null) 
                    {
                        averagingObservationsIDRef_ = averagingObservationsNode.Attributes["id"].Value;
                        AveragingObservationList ob = new AveragingObservationList(averagingObservationsNode);
                        IDManager.SetID(averagingObservationsIDRef_, ob);
                    }
                    else if (averagingObservationsNode.Attributes["href"] != null)
                    {
                        averagingObservationsIDRef_ = averagingObservationsNode.Attributes["href"].Value;
                    }
                    else
                    {
                        averagingObservations_ = new AveragingObservationList(averagingObservationsNode);
                    }
                }
                else
                {
                    averagingObservations_ = new AveragingObservationList(averagingObservationsNode);
                }
            }
            
        
            XmlNode marketDisruptionNode = xmlNode.SelectSingleNode("marketDisruption");
            
            if (marketDisruptionNode != null)
            {
                if (marketDisruptionNode.Attributes["href"] != null || marketDisruptionNode.Attributes["id"] != null) 
                {
                    if (marketDisruptionNode.Attributes["id"] != null) 
                    {
                        marketDisruptionIDRef_ = marketDisruptionNode.Attributes["id"].Value;
                        MarketDisruption ob = new MarketDisruption(marketDisruptionNode);
                        IDManager.SetID(marketDisruptionIDRef_, ob);
                    }
                    else if (marketDisruptionNode.Attributes["href"] != null)
                    {
                        marketDisruptionIDRef_ = marketDisruptionNode.Attributes["href"].Value;
                    }
                    else
                    {
                        marketDisruption_ = new MarketDisruption(marketDisruptionNode);
                    }
                }
                else
                {
                    marketDisruption_ = new MarketDisruption(marketDisruptionNode);
                }
            }
            
        
        }
        
    
        #region Schedule_
        private List<AveragingSchedule> schedule_;
        public List<AveragingSchedule> Schedule_
        {
            get
            {
                if (this.schedule_ != null)
                {
                    return this.schedule_; 
                }
                else if (this.scheduleIDRef_ != null)
                {
                    return this.schedule_; 
                }
                else
                {
                      return this.schedule_; 
                }
            }
            set
            {
                if (this.schedule_ != value)
                {
                    this.schedule_ = value;
                }
            }
        }
        #endregion
        
        public string scheduleIDRef_ { get; set; }
        #region AveragingDateTimes_
        private DateTimeList averagingDateTimes_;
        public DateTimeList AveragingDateTimes_
        {
            get
            {
                if (this.averagingDateTimes_ != null)
                {
                    return this.averagingDateTimes_; 
                }
                else if (this.averagingDateTimesIDRef_ != null)
                {
                    averagingDateTimes_ = IDManager.getID(averagingDateTimesIDRef_) as DateTimeList;
                    return this.averagingDateTimes_; 
                }
                else
                {
                      return this.averagingDateTimes_; 
                }
            }
            set
            {
                if (this.averagingDateTimes_ != value)
                {
                    this.averagingDateTimes_ = value;
                }
            }
        }
        #endregion
        
        public string averagingDateTimesIDRef_ { get; set; }
        #region AveragingObservations_
        private AveragingObservationList averagingObservations_;
        public AveragingObservationList AveragingObservations_
        {
            get
            {
                if (this.averagingObservations_ != null)
                {
                    return this.averagingObservations_; 
                }
                else if (this.averagingObservationsIDRef_ != null)
                {
                    averagingObservations_ = IDManager.getID(averagingObservationsIDRef_) as AveragingObservationList;
                    return this.averagingObservations_; 
                }
                else
                {
                      return this.averagingObservations_; 
                }
            }
            set
            {
                if (this.averagingObservations_ != value)
                {
                    this.averagingObservations_ = value;
                }
            }
        }
        #endregion
        
        public string averagingObservationsIDRef_ { get; set; }
        #region MarketDisruption_
        private MarketDisruption marketDisruption_;
        public MarketDisruption MarketDisruption_
        {
            get
            {
                if (this.marketDisruption_ != null)
                {
                    return this.marketDisruption_; 
                }
                else if (this.marketDisruptionIDRef_ != null)
                {
                    marketDisruption_ = IDManager.getID(marketDisruptionIDRef_) as MarketDisruption;
                    return this.marketDisruption_; 
                }
                else
                {
                      return this.marketDisruption_; 
                }
            }
            set
            {
                if (this.marketDisruption_ != value)
                {
                    this.marketDisruption_ = value;
                }
            }
        }
        #endregion
        
        public string marketDisruptionIDRef_ { get; set; }
        public string choiceStr_0;
        
    
        
    
    }
    
}

